File: client/web/src/features/webauthn/api/webauthn-api.ts
import type {
  PublicKeyCredentialCreationOptionsJSON,
  PublicKeyCredentialRequestOptionsJSON,
  RegistrationResponseJSON,
  AuthenticationResponseJSON,
} from "@simplewebauthn/browser";
import { apiClient } from "@/lib/api-client";
import type { WebAuthnCredentialInfo, Envelope } from "@/types/api-types";

// Registration (authenticated)
export function beginRegistration(): Promise<{
  ceremonyId: string;
  publicKeyOptions: PublicKeyCredentialCreationOptionsJSON;
}> {
  return apiClient.post("/webauthn/register/begin");
}

export function completeRegistration(
  ceremonyId: string,
  friendlyName: string,
  attestationResponse: RegistrationResponseJSON,
): Promise<{ id: string; friendlyName: string }> {
  return apiClient.post("/webauthn/register/complete", {
    ceremonyId,
    friendlyName,
    attestationResponse,
  });
}

// Assertion (public — login or recovery step-up)
export function beginAssertion(
  email?: string,
  recoverySessionId?: string,
): Promise<{
  ceremonyId: string;
  publicKeyOptions: PublicKeyCredentialRequestOptionsJSON;
}> {
  return apiClient.post("/webauthn/assert/begin", { email, recoverySessionId });
}

export function completeAssertion(
  ceremonyId: string,
  assertionResponse: AuthenticationResponseJSON,
  device?: { deviceName: string; platform: string; clientType: string; devicePublicKey: string },
): Promise<{
  success?: boolean;
  userId?: string;
  accessToken?: string | null;
  refreshToken?: string | null;
  device?: { deviceId: string; status: string; persistentKeyAllowed: boolean };
  envelopes?: { password?: Envelope; device?: Envelope };
  approvalRequestId?: string;
  role?: string;
  forcePasswordReset?: boolean;
}> {
  return apiClient.post("/webauthn/assert/complete", {
    ceremonyId,
    assertionResponse,
    device,
  });
}

// Management (authenticated)
export function listCredentials(): Promise<{ credentials: WebAuthnCredentialInfo[] }> {
  return apiClient.get("/webauthn/credentials");
}

export function renameCredential(credentialId: string, friendlyName: string): Promise<void> {
  return apiClient.put(`/webauthn/credentials/${credentialId}/rename`, { friendlyName });
}

export function revokeCredential(credentialId: string): Promise<void> {
  return apiClient.post(`/webauthn/credentials/${credentialId}/revoke`);
}

File: client/android/core/src/main/java/io/github/pstanar/pstotp/core/api/WebAuthnApi.kt
package io.github.pstanar.pstotp.core.api

import io.github.pstanar.pstotp.core.model.api.DeviceDto
import io.github.pstanar.pstotp.core.model.api.LoginCompleteResponse
import io.github.pstanar.pstotp.core.model.api.WebAuthnBeginResponse
import io.github.pstanar.pstotp.core.model.api.WebAuthnCredentialInfo
import org.json.JSONObject

/** WebAuthn/passkey API endpoints. */
class WebAuthnApi(private val client: ApiClient) {

    // --- Registration (authenticated) ---

    suspend fun beginRegistration(): WebAuthnBeginResponse =
        client.post("/webauthn/register/begin", null) { WebAuthnBeginResponse.fromJson(it) }

    suspend fun completeRegistration(
        ceremonyId: String,
        friendlyName: String,
        attestationResponseJson: String,
    ) {
        val body = JSONObject().apply {
            put("ceremonyId", ceremonyId)
            put("friendlyName", friendlyName)
            put("attestationResponse", JSONObject(attestationResponseJson))
        }
        client.post("/webauthn/register/complete", body) { }
    }

    // --- Assertion (public — login or recovery step-up) ---

    suspend fun beginAssertion(
        email: String? = null,
        recoverySessionId: String? = null,
    ): WebAuthnBeginResponse {
        val body = JSONObject().apply {
            if (email != null) put("email", email)
            if (recoverySessionId != null) put("recoverySessionId", recoverySessionId)
        }
        return client.postPublic("/webauthn/assert/begin", body) { WebAuthnBeginResponse.fromJson(it) }
    }

    suspend fun completeAssertionForLogin(
        ceremonyId: String,
        assertionResponseJson: String,
        device: DeviceDto,
    ): LoginCompleteResponse {
        val body = JSONObject().apply {
            put("ceremonyId", ceremonyId)
            put("assertionResponse", JSONObject(assertionResponseJson))
            put("device", device.toJson())
        }
        return client.postPublic("/webauthn/assert/complete", body) { LoginCompleteResponse.fromJson(it) }
    }

    suspend fun completeAssertionForRecovery(
        ceremonyId: String,
        assertionResponseJson: String,
    ) {
        val body = JSONObject().apply {
            put("ceremonyId", ceremonyId)
            put("assertionResponse", JSONObject(assertionResponseJson))
        }
        client.postPublic("/webauthn/assert/complete", body) { }
    }

    // --- Credential management (authenticated) ---

    suspend fun listCredentials(): List<WebAuthnCredentialInfo> =
        client.get("/webauthn/credentials") { json ->
            val arr = json.getJSONArray("credentials")
            (0 until arr.length()).map { WebAuthnCredentialInfo.fromJson(arr.getJSONObject(it)) }
        }

    suspend fun renameCredential(credentialId: String, friendlyName: String) {
        val body = JSONObject().put("friendlyName", friendlyName)
        client.put("/webauthn/credentials/$credentialId/rename", body) { }
    }

    suspend fun revokeCredential(credentialId: String) {
        client.post("/webauthn/credentials/$credentialId/revoke", null) { }
    }
}
